Fix connectors with custom templates in Helm Chart

weimerb-tng opened the pull request:
This enables connectors with custom templates to provide custom MS Teams message templates.

Fixes #164

Knappek commented on chart/prometheus-msteams/templates/configMapTemplate.yaml:
```diff
@@ -10,3 +10,11 @@ binaryData:
 {{- else }}
```
How about completely removing `.Values.customCardTemplate` in favor of `.Values.customCardTemplates`?

githubfellow87123 replied:
If it's okay to make a non backward compatible change here that's fine for me. So we'd go to version `0.8.0` then?

weimerb-tng replied:
If it's okay to make a non backward compatible change here that's fine for me. So we'd go to version `0.8.0` then?

Knappek replied:
Yeah why not, but as of my comments below, we would also remove `.Values.customCardTemplates`. As this is a breaking change, it would be rather version `1.0.0`

weimerb-tng replied:
Changed version to `1.0.0`

Knappek commented on chart/prometheus-msteams/values.yaml:
```diff
@@ -45,7 +45,22 @@ metrics:
+#   template_file: /etc/template/custom-message-card.tmpl
+#   webhook_url: <webhook>
+#   escape_underscores: true
 
```
In order for this to work, we need a config.yaml like this:

```yaml
customCardTemplates:
  - templateFileName: card1.tmpl
  - templateFileName: card2.tmpl

connectorsWithCustomTemplates:
  - request_path: /alert1
    template_file: /etc/template/card1.tmpl
    webhook_url: https://outlook.office.com/webhook/xxxx
    escape_underscores: true
  - request_path: /alert2
    template_file: /etc/template/card2.tmpl
    webhook_url: https://outlook.office.com/webhook/yyyy
    escape_underscores: true
```

What I don't like in this approach is, that you need to provide both `customCardTemplates` and `connectorsWithCustomTemplates`. Additionally with `template_file: /etc/template/card1.tmpl` the user always has to provide `/etc/template/` (this shouldn't be the users concern), and the `card1.tmpl` in  /etc/template/card1.tmpl` has to match the `- templateFileName: card1.tmpl` in `customCardTemplates`.

How about we make this more easy: what if the user simply provides:

```yaml
connectorsWithCustomTemplates:
  - request_path: /alert1
    webhook_url: https://outlook.office.com/webhook/xxxx
    escape_underscores: true
  - request_path: /alert2
    webhook_url: https://outlook.office.com/webhook/yyyy
    escape_underscores: true
```

where `template_file` then can be provided via `--set-file`, i.e. with

```shell
helm upgrade --install prometheus-msteams \
  --namespace default -f ~/.prometheus-msteams/helm-config.yaml \
  --set-file connectorsWithCustomTemplates\[0\].template_file=./default-message-card.tmpl \
  --set-file connectorsWithCustomTemplates\[1\].template_file=./examples/templates/card-with-action.tmpl \
  ./chart/prometheus-msteams
```

A second option is to provide `template_file` inline like

```yaml
connectorsWithCustomTemplates:
  - request_path: /alert1
    webhook_url: https://outlook.office.com/webhook/xxxx
    escape_underscores: true
    template_file: |
      {{ define "teams.card" }}
      ...
      {{ end }}
```

and then we completely get rid of `customCardTemplates`. What do you think?

githubfellow87123 replied:
Good point, this approach looks much cleaner.

weimerb-tng replied:
Good point, this approach looks much cleaner.

Knappek commented on chart/prometheus-msteams/values.yaml:
```diff
@@ -45,7 +45,22 @@ metrics:
+#   escape_underscores: true
 
 ## Specify the custom message card template for MS teams
 # customCardTemplate: |
```
see above, I think we can get rid of that.

githubfellow87123 replied:
I'll remove it

weimerb-tng replied:
I'll remove it

Knappek commented on chart/prometheus-msteams/README.md:
```diff
@@ -86,6 +87,26 @@ helm upgrade --install prometheus-msteams \
+### Customise Messages per MS Teams Channel
+
+To define a custom message template per MS Teams channel you can use the following configuration.
+Like descibed in [Customise messages to MS Teams](#customise-messages-to-ms-teams) you can use the `--set-file` flag
```
Please be more concise and provide an example how to use `--set-file` as it is not that easy when working with arrays (see my example in my review comment below).

githubfellow87123 replied:
That's true, I'll update this part to be more concise how to use --set-file with an array.

weimerb-tng replied:
That's true, I'll update this part to be more concise how to use --set-file with an array.

githubfellow87123 commented:
Hi @Knappek,

changed the chart according to your review comments. IMO there are some issues with these changes as well

1. A single custom card template can not be used anymore without specifying a connector with `connectorsWithCustomTemplates` (not that bad, since there is an other way to do it)
2. The spec of `connectorsWithCustomTemplates` in the helm chart has to be updated when the struct `ConnectorWithCustomTemplate` changes (can be annoying) 

https://github.com/prometheus-msteams/prometheus-msteams/blob/68ac740b84debc832c7dee7166ac06c3e11805c8/cmd/server/main.go#L47-L52 changes

Are these issues you can live with or do you see a way to improve this?

Kind regards
weimerb-tng

weimerb-tng commented:
Hi @Knappek,

changed the chart according to your review comments. IMO there are some issues with these changes as well

1. A single custom card template can not be used anymore without specifying a connector with `connectorsWithCustomTemplates` (not that bad, since there is an other way to do it)
2. The spec of `connectorsWithCustomTemplates` in the helm chart has to be updated when the struct `ConnectorWithCustomTemplate` changes (can be annoying) 

https://github.com/prometheus-msteams/prometheus-msteams/blob/68ac740b84debc832c7dee7166ac06c3e11805c8/cmd/server/main.go#L47-L52 changes

Are these issues something you can live with or do you see a way to improve this?

Kind regards
weimerb-tng

Knappek commented on chart/prometheus-msteams/README.md:
```diff
@@ -70,22 +70,40 @@ helm upgrade --install prometheus-msteams \
 helm upgrade --install prometheus-msteams \
   --namespace default -f config.yaml \
-  --set-file customCardTemplate=custom-card.tmpl \
+  "connectorsWithCustomTemplates[0].template_file=custom-card.tmpl" \
```
```suggestion
  --set-file "connectorsWithCustomTemplates[0].template_file=custom-card.tmpl" \
```

Knappek commented on chart/prometheus-msteams/templates/configMapTemplate.yaml:
```diff
@@ -5,8 +5,10 @@ metadata:
-{{ .Values.customCardTemplate | b64enc | indent 4 }}
-{{- else }}
 {{ .Files.Get "card.tmpl" | b64enc | indent 4 }}
+{{- range $index, $customCardTemplate := .Values.connectorsWithCustomTemplates }}
+{{- if hasKey $customCardTemplate "template_file" }}
+  custom_card_{{ $index }}.tmpl: |-
+{{ $customCardTemplate.template_file | b64enc | indent 4 }}
+{{- end }}
```
unfortunately this does not work when using `--set-file`. I was trying a little bit whether I can fix it, but don't have a solution yet. Somehow we need to do a dynamic `{{ .Files.Get <filename> | b64enc | indent 4 }}`.

weimerb-tng replied:
Hmm, works fine for me. This is how I executed it

```
helm template chart/prometheus-msteams/ -f values.yaml --set-file "connectorsWithCustomTemplates[0].template_file=custom-message-card.tmpl"
```

This is the content of `values.yaml`

```
connectorsWithCustomTemplates:
  - request_path: /alert2
    template_file:
    webhook_url: https://webhook
    escape_underscores: true
```

And this is the content of `custom-message-card.tmpl`

```
{{ define "teams.card" }}
{...}
{{ end }}
```

Output of the config map is

```
# Source: prometheus-msteams/templates/configMapTemplate.yaml
apiVersion: v1
kind: ConfigMap
metadata:
 name: prometheus-msteams-card-template
binaryData:
  card.tmpl: |-
    e3sgZGVmaW5lICJ0ZWFtcy5jYXJkIiB9fQp7CiAgIkB0eXBlIjogIk1lc3NhZ2VDYXJkIiwKICAiQGNvbnRleHQiOiAiaHR0cDovL3NjaGVtYS5vcmcvZXh0ZW5zaW9ucyIsCiAgInRoZW1lQ29sb3IiOiAie3stIGlmIGVxIC5TdGF0dXMgInJlc29sdmVkIiAtfX0yREM3MkQKICAgICAgICAgICAgICAgICB7ey0gZWxzZSBpZiBlcSAuU3RhdHVzICJmaXJpbmciIC19fQogICAgICAgICAgICAgICAgICAgIHt7LSBpZiBlcSAuQ29tbW9uTGFiZWxzLnNldmVyaXR5ICJjcml0aWNhbCIgLX19OEMxQTFBCiAgICAgICAgICAgICAgICAgICAge3stIGVsc2UgaWYgZXEgLkNvbW1vbkxhYmVscy5zZXZlcml0eSAid2FybmluZyIgLX19RkZBNTAwCiAgICAgICAgICAgICAgICAgICAge3stIGVsc2UgLX19ODA4MDgwe3stIGVuZCAtfX0KICAgICAgICAgICAgICAgICB7ey0gZWxzZSAtfX04MDgwODB7ey0gZW5kIC19fSIsCiAgInN1bW1hcnkiOiAie3stIGlmIGVxIC5Db21tb25Bbm5vdGF0aW9ucy5zdW1tYXJ5ICIiIC19fQogICAgICAgICAgICAgICAgICB7ey0gaWYgZXEgLkNvbW1vbkFubm90YXRpb25zLm1lc3NhZ2UgIiIgLX19CiAgICAgICAgICAgICAgICAgICAge3stIGlmIGVxIC5Db21tb25MYWJlbHMuYWxlcnRuYW1lICIiIC19fQogICAgICAgICAgICAgICAgICAgICAgUHJvbWV0aGV1cyBBbGVydAogICAgICAgICAgICAgICAgICAgIHt7LSBlbHNlIC19fQogICAgICAgICAgICAgICAgICAgICAge3stIC5Db21tb25MYWJlbHMuYWxlcnRuYW1lIC19fQogICAgICAgICAgICAgICAgICAgIHt7LSBlbmQgLX19CiAgICAgICAgICAgICAgICAgIHt7LSBlbHNlIC19fQogICAgICAgICAgICAgICAgICAgIHt7LSAuQ29tbW9uQW5ub3RhdGlvbnMubWVzc2FnZSAtfX0KICAgICAgICAgICAgICAgICAge3stIGVuZCAtfX0KICAgICAgICAgICAgICB7ey0gZWxzZSAtfX0KICAgICAgICAgICAgICAgICAge3stIC5Db21tb25Bbm5vdGF0aW9ucy5zdW1tYXJ5IC19fQogICAgICAgICAgICAgIHt7LSBlbmQgLX19IiwKICAidGl0bGUiOiAiUHJvbWV0aGV1cyBBbGVydCAoe3sgLlN0YXR1cyB8IHRpdGxlIH19KSIsCiAgInNlY3Rpb25zIjogWyB7eyRleHRlcm5hbFVybCA6PSAuRXh0ZXJuYWxVUkx9fQogIHt7LSByYW5nZSAkaW5kZXgsICRhbGVydCA6PSAuQWxlcnRzIH19e3stIGlmICRpbmRleCB9fSx7ey0gZW5kIH19CiAgICB7CiAgICAgICJhY3Rpdml0eVRpdGxlIjogIlt7eyAkYWxlcnQuQW5ub3RhdGlvbnMuZGVzY3JpcHRpb24gfX1dKHt7ICRleHRlcm5hbFVybCB9fSkiLAogICAgICAiZmFjdHMiOiBbCiAgICAgICAge3stIHJhbmdlICRrZXksICR2YWx1ZSA6PSAkYWxlcnQuQW5ub3RhdGlvbnMgfX0KICAgICAgICB7CiAgICAgICAgICB7ey0gaWYgbmUgJGtleSAiZGVzY3JpcHRpb24iIC19fQogICAgICAgICAgICAibmFtZSI6ICJ7eyAka2V5IH19IiwKICAgICAgICAgICAgInZhbHVlIjogInt7ICR2YWx1ZSB9fSIKICAgICAgICAgIHt7LSBlbmQgLX19CiAgICAgICAgfSwKICAgICAgICB7ey0gZW5kIC19fQogICAgICAgIHt7JGMgOj0gY291bnRlcn19e3sgcmFuZ2UgJGtleSwgJHZhbHVlIDo9ICRhbGVydC5MYWJlbHMgfX17e2lmIGNhbGwgJGN9fSx7eyBlbmQgfX0KICAgICAgICB7CiAgICAgICAgICAibmFtZSI6ICJ7eyAka2V5IH19IiwKICAgICAgICAgICJ2YWx1ZSI6ICJ7eyAkdmFsdWUgfX0iCiAgICAgICAgfQogICAgICAgIHt7LSBlbmQgfX0KICAgICAgXSwKICAgICAgIm1hcmtkb3duIjogdHJ1ZQogICAgfQogICAge3stIGVuZCB9fQogIF0KfQp7eyBlbmQgfX0K
  custom_card_0.tmpl: |-
    e3sgZGVmaW5lICJ0ZWFtcy5jYXJkIiB9fQp7Li4ufQp7eyBlbmQgfX0K
```

Knappek replied:
late evening typo *facepalm*. Awesome, this works great! :). I really like that, thanks a lot @weimerb-tng!

Knappek commented on chart/prometheus-msteams/Chart.yaml:
```diff
@@ -3,7 +3,7 @@ appVersion: "v1.4.1"
 name: prometheus-msteams
 home: https://github.com/prometheus-msteams/prometheus-msteams
-version: 0.7.1
+version: 0.8.0
```
see my comment from previous review: this would be version `1.0.0` when we introduce a breaking change.

Knappek commented:
> Hi @Knappek,
> 
> changed the chart according to your review comments. IMO there are some issues with these changes as well
> 
> 1. A single custom card template can not be used anymore without specifying a connector with `connectorsWithCustomTemplates` (not that bad, since there is an other way to do it)
> 2. The spec of `connectorsWithCustomTemplates` in the helm chart has to be updated when the struct `ConnectorWithCustomTemplate` changes (can be annoying)
> 3. Travis build failed, not sure why.
> 
> https://github.com/prometheus-msteams/prometheus-msteams/blob/68ac740b84debc832c7dee7166ac06c3e11805c8/cmd/server/main.go#L47-L52
> 
> changes
> Are these issues something you can live with or do you see a way to improve this?
> 
> Kind regards
> weimerb-tng

@weimerb-tng thank you for your work and opinion.  I added some review comments above. Additionally:

1. Hmm you are right. Should we still continue to use `customCardTemplate` which would replace the default card template? For anything else we won't use `customCardTemplates` anymore, so we deprecated that field in favor of `connectorsWithCustomTemplates`. Having both would be fine for me.
2. Yes you are right, but I think it is fine (for now). 
3. seems like there was a connection issue in travis, I restarted the build, let's see whether it works now.

weimerb-tng commented:
@Knappek Thanks for the quick response.

1. I reverted the change to remove `customCardTemplate`. So it's still possbile to just set a custom MS Teams Card without specifying additional connectors. The variable `customCardTemplates` was introduced in this Pull Request. So there is no need to deprecate it or smth. With this changeset we should't have any backward incompatible changes anymore right? So I changed the helm chart version back to `0.7.2`.

Knappek commented on chart/prometheus-msteams/README.md:
```diff
@@ -86,6 +87,40 @@ helm upgrade --install prometheus-msteams \
+### Customise Messages per MS Teams Channel
+
+This application uses a [Default Teams Message Card Template](./prometheus-msteams/card.tmpl) to convert incoming Prometheus alerts to teams message cards.
+To define a custom message template per MS Teams channel you can use the following configuration.
```
```suggestion
To define a custom message template per MS Teams channel you can use the following configuration:
```

Knappek commented on chart/prometheus-msteams/README.md:
````diff
@@ -86,6 +87,40 @@ helm upgrade --install prometheus-msteams \
+```yaml
+connectorsWithCustomTemplates:
+  - request_path: /alert2
+    webhook_url: <webhook>
````
```suggestion
    webhook_url: https://outlook.office.com/webhook/xxxx/xxxx
```

Knappek replied:
I'd rather prefer having this for consistency with the rest of the readme.

Knappek commented on chart/prometheus-msteams/README.md:
````diff
@@ -86,6 +87,40 @@ helm upgrade --install prometheus-msteams \
+```yaml
+connectorsWithCustomTemplates:
+  - request_path: /alert2
+    webhook_url: <webhook>
````
```suggestion
    webhook_url: https://outlook.office.com/webhook/xxxx/xxxx
```

Knappek commented on chart/prometheus-msteams/README.md:
````diff
@@ -86,6 +87,40 @@ helm upgrade --install prometheus-msteams \
+      {{ end }}
+```
 
 ### Helm Configuration
````
This table still displays `connectors_with_custom_templates` instead of `connectorsWithCustomTemplates`. I was thinking: does this naming change introduce a major change, although it is actually a fix as it didn't work before? What do you think?

weimerb-tng replied:
I guess you are right. You can not use the old name anymore, so it should be a major change. Shall we update the version to `1.0.0` then?

Knappek replied:
Yes let's do it. I can't directly link the line in the readme, this is what I meant:
![image](https://user-images.githubusercontent.com/18685317/100153668-b7e9d000-2ea4-11eb-8756-c3b7b233cabe.png)

See that `connectors_with_custom_templates`?

weimerb-tng replied:
Ahh right, changed it and updated the chart version.

Knappek commented:
Awesome. Thank you very much for this contribution @weimerb-tng .
